src/Domain/Services/ExtendRestApi.php

```diff
@@ -0,0 +1,131 @@
+<?php
+namespace Automattic\WooCommerce\Blocks\Domain\Services;
+
+use Throwable;
+
+/**
+ * Service class to provide utility functions to extend REST API.
+ */
+class ExtendRestApi {
+	/**
+	 * Valid endpoints to extend
+	 *
+	 * @var array
+	 */
+	private $endpoints = [ 'item' ];
+
+	/**
+	 * An endpoint that validates registration method call
+	 *
+	 * @param string   $endpoint The endpoint to extend.
+	 * @param string   $namespace Plugin namespace.
+	 * @param callable $schema_callback Callback executed to add schema data.
+	 * @param callable $data_callback Callback executed to add endpoint data.
+	 */
+	private function validate_endpoint_data( $endpoint, $namespace, $schema_callback, $data_callback ) {
+		if ( ! is_string( $namespace ) ) {
+			$this->throw_exception(
+				'You must provide a plugin namespace when extending a Store REST endpoint.'
+			);
+		}
+
+		if ( ! is_string( $endpoint ) || ! in_array( $endpoint, $this->endpoints, true ) ) {
+			$this->throw_exception(
+				'You must provide a valid Store REST endpoint to extend, valid endpoints are: ' . implode( ', ', $this->endpoints )
+			);
+		}
+
+		if ( ! is_callable( $schema_callback ) ) {
+			$this->throw_exception(
+				'$schema_callback must be a callable function.'
+			);
+		}
+
+		if ( ! is_callable( $data_callback ) ) {
+			$this->throw_exception(
+				'$data_callback must be a callable function.'
+			);
+		}
+	}
+
+	/**
+	 * Extends the items in store/cart endpoint
+	 *
+	 * @param string   $namespace Plugin namespace.
+	 * @param callable $schema_callback Callback executed to add schema data.
+	 * @param callable $data_callback Callback executed to add endpoint data, gets cart item data as param.
+	 */
+	public function register_cart_item_endpoint_data( $namespace, $schema_callback, $data_callback ) {
+		$this->validate_endpoint_data( 'item', $namespace, $schema_callback, $data_callback );
+
+		add_filter(
+			'__internal_extend_cart_item_schema',
+			function( $schema ) use ( $schema_callback, $namespace ) {
+				$schema_data = [];
+
+				try {
+					$schema_data = $schema_callback();
+				} catch ( Throwable $e ) {
+					$this->throw_exception( $e );
+					return $schema;
+				}
+
+				if ( ! is_array( $schema_data ) ) {
+					$this->throw_exception(
+						'$schema_callback must return an array.'
+					);
+					return $schema;
+				}
+
+				$schema[ $namespace ] = $schema_data;
```

Comment:
Each endpoint data is going to be added to that namespace.

Reply:
When you implement what I suggested earlier about registering to an identifier, you would have something like:

```php
$schema[ $identifier ][ $namespace ] = $schema_data;
```

Then the getter that `AbstractRoute` or `AbstractSchema` would implement from `ExentdRestApi` should be able to retrieve all the registered extra data via the identifier for the concrete route/schema.